plugins/outputs/cloud_pubsub/pubsub.go

```diff
@@ -220,8 +237,18 @@ func (ps *PubSub) toMessages(metrics []telegraf.Metric) ([]*pubsub.Message, erro
 			encoded := base64.StdEncoding.EncodeToString(b)
 			b = []byte(encoded)
 		}
-
-		msg := &pubsub.Message{Data: b}
+		b, err = ps.encoder.Encode(b)
+		if err != nil {
+			return nil, err
+		}
+		var data []byte
+		if ps.ContentEncoding == "gzip" {
+			data = make([]byte, len(b))
+			copy(data, b)
+		} else {
+			data = b
+		}
```

Comment:
IIRC `Write()` is not called concurrently, so `toMessages()` isn't either. Therefore, I think you do not need to copy the encoded buffer here.